File: backend/backend/agent/controller.py
"""Agentic controller (backend/agent/controller.py) — the central
orchestration state machine.

Responsibilities: receive validated input + structured intent, select a
workflow/model from the registry, execute the corresponding tool, and
record an observable trace at every stage. This mirrors
src/server/satquery/controller.ts in the live Next.js implementation.
"""
from __future__ import annotations

from typing import Any, Optional

from backend.query_understanding.router import classify_intent
from backend.agent.registry import load_registry, resolve_model
from backend.output.trace_logger import TraceLogger
from backend.tools.vqa_tool import run_vqa
from backend.tools.caption_tool import run_caption
from backend.tools.grounding_tool import run_grounding
from backend.tools.change_tool import run_change
from backend.tools.optical_sar_tool import run_optical_sar
from backend.evidence.evidence_engine import build_evidence
from backend.evidence.confidence import compute_confidence


class AgentController:
    def __init__(self) -> None:
        self.registry = load_registry()

    def orchestrate(self, query: str, mode: Optional[str], files: list[dict], trace: TraceLogger) -> dict[str, Any]:
        trace.log("query_understanding", "Classifying natural-language query", {"query": query, "mode": mode})
        intent = classify_intent(query, mode)
        trace.log("query_understanding", f'Resolved task="{intent.task}" target="{intent.target}"', {})
        if intent.coerced:
            trace.log("query_understanding", f"Coerced: {intent.coercion_reason}", {})

        entry = self.registry[intent.task]
        trace.log("routing", f'Selected workflow "{entry.tool}" for task "{intent.task}"', {})

        primary = resolve_model(entry.primary_name)
        models_used = []
        used_fallback = False
        if primary is not None:
            health = primary.health_check()
        else:
            health = {"available": False, "reason": f"{entry.primary_name} has no local wrapper class configured."}

        if health["available"]:
            models_used.append({"name": primary.name, "role": primary.role, "status": "used"})
        else:
            models_used.append({"name": entry.primary_name, "role": entry.tool, "status": "unavailable", "reason": health.get("reason")})
            used_fallback = True
            if entry.fallback_name:
                models_used.append({"name": entry.fallback_name, "role": "fallback", "status": "fallback_used"})
                trace.log("fallback", f'Falling back to "{entry.fallback_name}" per registry policy.', {})

        trace.log("execution", f'Executing tool "{entry.tool}"', {})

        target = intent.target
        tool_output = self._execute_tool(intent.task, query, files, target, entry)

        trace.log("execution", f'Tool "{entry.tool}" completed.', {})

        evidence = build_evidence(tool_output.get("facts", []), tool_output, tool_output.get("tool_agreement", 1.0), tool_output.get("model_probability", 0.6))
        confidence = compute_confidence(tool_output.get("model_probability", 0.6), tool_output.get("tool_agreement", 1.0), 1.0)
        trace.log("evidence", "Built structured evidence.", {})
        trace.log("confidence", f"Confidence estimate: {confidence['confidence_estimate']}", {})

        return {
            "mode": mode,
            "intent": intent.__dict__,
            "workflow": entry.tool,
            "models_used": models_used,
            "used_fallback": used_fallback,
            "evidence": evidence,
            "confidence": confidence,
            "answer": tool_output.get("answer"),
            "warnings": [],
        }

    def _execute_tool(self, task: str, query: str, files: list[dict], target: Optional[str], entry) -> dict:
        # NOTE: this reference implementation focuses on interface fidelity.
        # Real raster loading (rasterio -> numpy arrays) is implemented in
        # gis/processor.py; wiring it end-to-end requires actual GeoTIFF
        # fixtures + rasterio/GDAL installed. The fully worked, end-to-end
        # numeric pipeline is implemented in the live Next.js app under
        # src/server/satquery/{tools,gis,controller}.ts.
        if task == "vqa":
            from backend.gis.processor import read_bands

            image_path = files[0]["stored_path"]

            try:
                data, profile = read_bands(image_path)

                band_stats = {
                    "width": int(data.shape[2]),
                    "height": int(data.shape[1]),
                    "band_count": int(data.shape[0]),
                    "target_coverage_percent": 0.0,
                }

                result = run_vqa(
                    image_path,
                    query,
                    target,
                    band_stats=band_stats,
                )

                result["raster"] = {
                    "width": int(data.shape[2]),
                    "height": int(data.shape[1]),
                    "band_count": int(data.shape[0]),
                    "crs": str(profile.get("crs")) if profile.get("crs") else None,
                }

                return result

            except Exception as exc:
                result = run_vqa(
                    image_path,
                    query,
                    target,
                    band_stats={"target_coverage_percent": 0.0},
                )

                result.setdefault("warnings", []).append(
                    f"Raster processing unavailable: {exc}"
                )

                return result
        if task == "caption":
            return run_caption(files[0]["stored_path"], composition={})
        if task == "grounding":
            import numpy as np

            return run_grounding(files[0]["stored_path"], target or "built_up", index_array=np.zeros((1, 1)))
        if task == "change_vqa":
            import numpy as np

            return run_change(files[0]["stored_path"], files[1]["stored_path"], query, target, np.zeros((1, 1)), np.zeros((1, 1)), res={})
        if task == "optical_sar":
            import numpy as np

            return run_optical_sar(files[0]["stored_path"], files[1]["stored_path"], target, np.zeros((1, 1)), np.zeros((1, 1)), res={})
        raise ValueError(f"Unhandled task: {task}")

File: frontend/src/server/satquery/controller.ts
// Agentic controller (backend/agent/controller.py) — the central
// orchestration state machine. It receives validated input + structured
// intent, selects a workflow/model from the registry, executes the
// corresponding tool, and records an observable trace at every stage.
import { promises as fs } from "fs";
import { classifyIntent } from "./queryUnderstanding";
import { resolveRegistry, validateParameters } from "./registry";
import { readRaster, type RasterData } from "./gis";
import { runCaptionTool, runChangeTool, runGroundingTool, runOpticalSarTool, runVqaTool, type ToolOutput } from "./tools";
import { predictRsCoVLM, RSCoVLM } from "./models";
import { TraceLogger } from "./trace";
import type { InputMode, ModelUsed, QueryIntent, UploadedFileMeta } from "./types";

export interface ControllerInput {
  query: string;
  mode: InputMode;
  files: UploadedFileMeta[];
}

export interface ControllerOutput {
  intent: QueryIntent;
  workflow: string;
  modelsUsed: ModelUsed[];
  parameters: Record<string, unknown>;
  toolOutput: ToolOutput;
  usedFallback: boolean;
}

async function loadRaster(file: UploadedFileMeta): Promise<RasterData> {
  const buffer = await fs.readFile(file.storedPath);
  const format = file.format === "unknown" ? "PNG" : file.format;
  return readRaster(buffer, format);
}

function fileByRole(files: UploadedFileMeta[], roles: string[]): UploadedFileMeta | undefined {
  return files.find((f) => roles.includes(f.role));
}

export async function orchestrate(input: ControllerInput, trace: TraceLogger): Promise<ControllerOutput> {
  trace.log("query_understanding", "Classifying natural-language query into a structured intent", {
    query: input.query,
    mode: input.mode,
  });
  const intent = await classifyIntent(input.query, input.mode);
  trace.log("query_understanding", `Resolved task="${intent.task}" target="${intent.target ?? "none"}" via ${intent.method}`, {
    intent,
  });
  if (intent.coerced) {
    trace.log("query_understanding", `Coerced task for input mode: ${intent.coercionReason}`, {});
  }

  const entry = resolveRegistry(intent.task);
  trace.log("routing", `Selected workflow/tool "${entry.tool}" for task "${entry.task}"`, { tool: entry.tool });

  const parameters: Record<string, unknown> = intent.target ? { target: intent.target } : {};
  const paramWarnings = validateParameters(intent.task, parameters);
  paramWarnings.forEach((w) => trace.log("routing", w, {}));

  const modelsUsed: ModelUsed[] = [];
  let usedFallback = false;

  const primaryHealth = await entry.primary.healthCheck();
  if (primaryHealth.available) {
    modelsUsed.push({ name: entry.primary.name, role: entry.primary.role, status: "used" });
    trace.log("model_selection", `Primary model "${entry.primary.name}" is available.`, {});
  } else {
    modelsUsed.push({ name: entry.primary.name, role: entry.primary.role, status: "unavailable", reason: primaryHealth.reason });
    trace.log("model_selection", `Primary model "${entry.primary.name}" unavailable: ${primaryHealth.reason}`, {});
    usedFallback = true;
    if (entry.fallback) {
      modelsUsed.push({ name: entry.fallback.name, role: entry.fallback.role, status: "fallback_used" });
      trace.log("fallback", `Falling back to configured fallback "${entry.fallback.name}" per registry policy.`, {});
    } else {
      trace.log("fallback", "No fallback configured for this task; request would fail without a rule-based tool implementation.", {});
    }
  }

  trace.log("execution", `Executing tool "${entry.tool}"`, { parameters });

  let toolOutput: ToolOutput;
  const target = (parameters.target as string | undefined) ?? null;

  switch (intent.task) {
    case "vqa": {
      const file = input.files[0];
      const raster = await loadRaster(file);
      toolOutput = runVqaTool(raster, input.query, target);
      break;
    }
    case "caption": {
      const file = input.files[0];
      const raster = await loadRaster(file);
      toolOutput = runCaptionTool(raster);
      break;
    }
    case "grounding": {
      const file = input.files[0];
      const raster = await loadRaster(file);
      toolOutput = runGroundingTool(raster, target ?? "built_up");
      break;
    }
    case "change_vqa": {
      const t1File = fileByRole(input.files, ["t1"]) ?? input.files[0];
      const t2File = fileByRole(input.files, ["t2"]) ?? input.files[1];
      const [t1, t2] = await Promise.all([loadRaster(t1File), loadRaster(t2File)]);
      toolOutput = runChangeTool(t1, t2, target);
      break;
    }
    case "optical_sar": {
      const opticalFile = fileByRole(input.files, ["optical"]) ?? input.files.find((f) => f.modality !== "sar") ?? input.files[0];
      const sarFile = fileByRole(input.files, ["sar"]) ?? input.files.find((f) => f.modality === "sar") ?? input.files[1];
      const [optical, sar] = await Promise.all([loadRaster(opticalFile), loadRaster(sarFile)]);
      toolOutput = runOpticalSarTool(optical, sar, target);
      break;
    }
    default:
      throw new Error(`Unhandled task type: ${intent.task}`);
  }

  // Keep the deterministic GIS evidence, overlays, and statistics, but replace
  // the user-facing language answer with the real EC2 GPU model response for
  // single-image tasks when RSCoVLM health succeeded above.
  if (!usedFallback && entry.primary === RSCoVLM && ["vqa", "caption", "grounding"].includes(intent.task)) {
    try {
      const file = input.files[0];
      const question = intent.task === "caption"
        ? input.query || "Describe this satellite image."
        : intent.task === "grounding"
          ? `Locate ${target ?? "the main target"} in this satellite image.`
          : input.query;
      const remote = await predictRsCoVLM(file.storedPath, question, intent.task);
      toolOutput = { ...toolOutput, answer: remote.answer || toolOutput.answer, modelProbability: 0.95 };
      trace.log("execution", `Remote RSCoVLM prediction completed on EC2 (${remote.model}).`, {});
    } catch (error) {
      usedFallback = true;
      trace.log("fallback", `Remote RSCoVLM prediction failed; retaining deterministic evidence answer: ${error instanceof Error ? error.message : String(error)}`, {});
    }
  }

  trace.log("execution", `Tool "${entry.tool}" completed successfully.`, {
    factCount: toolOutput.facts.length,
    toolAgreement: toolOutput.toolAgreement,
    modelProbability: toolOutput.modelProbability,
  });

  return { intent, workflow: entry.tool, modelsUsed, parameters, toolOutput, usedFallback };
}
